mater/lib/src/file_reader.rs

```diff
@@ -0,0 +1,246 @@
+{
+    reader: v2::Reader<R>,
+    index: HashMap<Cid, PartialNode>,
+}
+
+impl FileLoader<File> {
+    /// Creates a [`FileLoader`] from the given file path.
+    pub async fn from_path<P>(path: P) -> Result<Self, Error>
+    where
+        P: AsRef<Path>,
+    {
+        let file = File::open(path).await?;
+        let mut loader = Self {
+            reader: v2::Reader::new(file),
+            index: HashMap::new(),
+        };
+        loader.index().await?;
+        Ok(loader)
+    }
+}
+
+impl FileLoader<Cursor<Vec<u8>>> {
+    /// Creates a [`FileLoader`] from a vector of bytes.
+    pub async fn from_vec(vec: Vec<u8>) -> Result<Self, Error> {
+        let mut loader = Self {
+            reader: v2::Reader::new(Cursor::new(vec)),
+            index: HashMap::new(),
+        };
+        loader.index().await?;
+        Ok(loader)
+    }
+}
+
+impl<R> FileLoader<R>
+where
+    R: AsyncRead + AsyncSeek + Unpin,
+{
+    /// Returns the root of the CAR file.
+    ///
+    /// If the number of roots is not 1, returns [`Error::WrongNumberOfRoots`].
+    pub async fn root(&mut self) -> Result<Cid, Error> {
+        self.reader.get_inner_mut().rewind().await?;
+        self.reader.read_pragma().await?;
+        self.reader.read_header().await?;
+        // We assume there's a single root
+        self.reader
+            .read_v1_header()
+            .await?
+            .roots
+            .first()
+            .copied()
+            .ok_or(Error::WrongNumberOfRoots)
+    }
+
+    /// Indexes the file.
+    ///
+    /// *Always* seeks to the start of the file before proceeding with indexing.
+    /// This function performs indexing *naively*, it will sequentially read all block "headers"
+    /// (Cid, data start offset and data length) without loading the blocks into memory.
+    async fn index(&mut self) -> Result<(), Error> {
+        // Indexing must always be made from the start
+        self.reader.get_inner_mut().rewind().await?;
+        let _ = self.reader.read_pragma().await?;
+        let v2_header = self.reader.read_header().await?;
+        let _ = self.reader.read_v1_header().await?;
+
+        let data_end = v2_header.data_offset + v2_header.data_size;
+
+        loop {
+            let position = self.reader.get_inner_mut().stream_position().await?;
+            if position >= data_end {
+                break;
+            }
+
+            let block_metadata = self.reader.read_block_metadata().await?;
+            self.index
+                .insert(block_metadata.cid, block_metadata.try_into()?);
+        }
+
+        Ok(())
+    }
+
+    /// Traverse the tree under the given [`Cid`], yielding the data in the leaves.
+    fn load_cid<'a>(
+        &'a mut self,
+        cid: &'a Cid,
+    ) -> impl Stream<Item = Result<(Cid, Vec<u8>), Error>> + 'a {
+        try_stream! {
+            let partial_node = self.index.get(&cid).ok_or(Error::InvalidCid)?;
+            let mut stack = VecDeque::new();
```

Comment:
this is not a stack, it's a queue. 
You're pushing back and popping from front.
This is BFS.
```suggestion
            let mut queue = VecDeque::new();
```

Reply:
Great catch, this was a stack until I realized my mistake in using said stack